### PR TITLE
Fix #2314: Add fast paths to BigDecimal.valueExact.

### LINKED ISSUE #2314: Infinite loop in BigDecimal.isValidLong
The following causes an infinite loop in SJS:
```scala
BigDecimal("9.223372E+285625056").isValidLong
```

On JVM, it correctly returns `false`:
```
Welcome to Scala 2.11.8 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_77).
Type in expressions for evaluation. Or try :help.

scala> BigDecimal("9.223372E+285625056").isValidLong
res0: Boolean = false
```

### PR DESCRIPTION
**Pull request by nicolasstucki**
Review by @sjrd.

### CODE REVIEW

File: javalib/src/main/scala/java/math/BigDecimal.scala [hunk 1]
```diff
@@ -1727,11 +1727,24 @@ class BigDecimal() extends Number with Comparable[BigDecimal] {
    *          number don't fit in the primitive type
    */
   private def valueExact(bitLengthOfType: Int): Long = {
-    val bigInteger = toBigIntegerExact()
-    if (bigInteger.bitLength() < bitLengthOfType)
-      bigInteger.longValue()
-    else
+    // Fast path to avoid some large BigInteger creations by toBigIntegerExact
+    val length10 =_scale - _precision
```
Comment (sjrd) [inline javalib/src/main/scala/java/math/BigDecimal.scala]: Missing space after `=`

File: javalib/src/main/scala/java/math/BigDecimal.scala [hunk 2]
```diff
@@ -1727,11 +1727,24 @@ class BigDecimal() extends Number with Comparable[BigDecimal] {
...
+    if (length10 < 0) {
+      // It is not an integral number
+      throw new ArithmeticException("Rounding necessary")
+    } else if (length10 > bitLengthOfType * 3) {
+      /* length10 is the digit length in base 10 of the integral
+       * number which is less or equal to it's length in octal representation
+       * (i.e. bitLengthOfType * 3)
+       */
       throw new ArithmeticException("Rounding necessary")
```
Comment (sjrd) [inline javalib/src/main/scala/java/math/BigDecimal.scala]: I don't understand this part, and hence I am not convinced it works.

Why not simply use `length10 > 19`, which is `Long.MaxValue.toString.length`, i.e., the largest number of base 10 digits that a `Long` can ever have?

Comment (sjrd) [inline javalib/src/main/scala/java/math/BigDecimal.scala]: Or is it `_scale > 19`?

File: javalib/src/main/scala/java/math/BigDecimal.scala [hunk 3]
```diff
@@ -1727,11 +1727,24 @@ class BigDecimal() extends Number with Comparable[BigDecimal] {
...
-    val bigInteger = toBigIntegerExact()
-    if (bigInteger.bitLength() < bitLengthOfType)
-      bigInteger.longValue()
-    else
+    // Fast path to avoid some large BigInteger creations by toBigIntegerExact
+    val length10 =_scale - _precision
+    if (length10 < 0) {
+      // It is not an integral number
+      throw new ArithmeticException("Rounding necessary")
+    } else if (length10 > bitLengthOfType * 3) {
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/math/BigDecimal.scala]: Should be `bitLengthOfType / 3`.

File: javalib/src/main/scala/java/math/BigDecimal.scala (hunk 2 above)
Comment (nicolasstucki) [inline javalib/src/main/scala/java/math/BigDecimal.scala]: Yes, `length10 > 19` should work fine.

**Comment by sjrd** [comment]
LGTM

**Pull request merged**